src/main/java/com/mongodb/hibernate/translate/attachment/AttachmentKey.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright 2024-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.hibernate.translate.attachment;
+
+/**
+ * A type safe attachment category enum.
+ *
+ * <p>It is used as the common contract between coordinating attachment producer (aka attacher) and consumer (aka
+ * detacher). To ensure safe attachment transferring, both parties need to present the identical {@link AttachmentKey}.
+ *
+ * <p>Note that Java {@code Enum} doesn't support generics type. This interface combines both {@code Enum} and type
+ * safety together.
+ *
+ * @param <T> generics type
+ * @see Attachment#attach(AttachmentKey, Object)
+ * @see Attachment#expect(AttachmentKey, Runnable)
+ */
+public interface AttachmentKey<T> {
+    AttachmentKey<String> COLUMN_NAME = new AttachmentKey<>() {};
+    AttachmentKey<String> COLLECTION_NAME = new AttachmentKey<>() {};
```

Comment:
`toString` of such objects does not seem to be descriptive, yet it is used in `Attachment.expect` (
`throw new IllegalArgumentException(format("No value attached for attachment key: %s", attachmentKey))`). Let's improve that.

Reply:
Good catch! Ideally the output not only includes key name, but also the type info.

Given the keys will be added incrementally, it would be great to make the addition as easy as possible (without manual labour overhead).

I come up with the following idea to output both generic type info and names using Java reflection:

```
    default String render() {
        return Arrays.stream(AttachmentKey.class.getDeclaredFields())
                .filter(field -> {
                    try {
                        return field.get(null) == this;
                    } catch (IllegalAccessException e) {
                        throw new RuntimeException(e);
                    }
                })
                .findFirst()
                .map(field -> format("%s: %s", field.getGenericType(), field.getName()))
                .orElse(toString());
    }
```

so for the following new key constant:
```
AttachmentKey<List<Map<String, Object>>> NO_NAME = new AttachmentKey<>() {};
```

the above method will output the following info:
```
AttachmentKey<java.util.List<java.util.Map<java.lang.String, java.lang.Object>>>: NO_NAME
```
Not sure whether it is over-engineering or not.

In normal visitor usage scenario, usually type matching is enough. It is rare to return multiple different result from the same visitxxx(..) method for if that is the case, a better design is to split it into multiple different visitor methods instead.